### PR TITLE
HAAR 1902 - Data layer

### PR DESCRIPTION
**Pull request by thomasridd**
### What
- Data layer methods for the new UI
- Home page

### How to review
This is a large PR but shouldn't be too difficult once the architecture is understood. A quick explanations

#### Data layer

Our data layer architecture is all changes made in these directories
`/server/data`
`/server/interfaces`
`/server/mappers`
`/server/services`
`/server/testutils`
`/server/utils`

There are three classes of object

Model objects: `BaseClient`, `Client`, `ClientSecrets`
Response objects: `GetBaseClientResponse`, `ListBaseClientsResponse`, ...
Request objects: `AddBaseClientRequest`, `UpdateBaseClientRequest`, ...

Functions have been created in `/server/mappers` to map Model objects to Requests and Responses back to Model objects. The motivation for this is to allow our backend to evolve independently. When the API changes (and we are expecting change) we should be able to adjust our mapping functions and continue.

So the majority of added functionality is to take our backend API and for each restful method ...,

1. Add a method to an **api client** that takes a Request, does a network call, then returns a Response (see `/data/baseClientApiClient.ts`)
_Test - Use `nock` to mock the network - affirm that the baseClientApi method makes the correct network call and returns the right data_

2. Add **mapper** methods that convert model objects into requests and responses into model objects 
_Test - Not implemented - on hold until we are happy with the API stability_

3. Add a method to a **service** that takes a token and model objects as arguments. It will instantiate one or more api clients using the token then perform an operation using api methods. 
_Test - Mock api client methods - affirm that the method instantiates the necessary clients using the token - affirm it returns the mapped response back as a Model object_

Error handling happens at controller level so tests are happy path - for now at least

#### Home page
All file changes not in the above folders relate to making a homepage that demonstrates this approach is working, and fits with the MoJ template architecture

### CODE REVIEW

File: server/config.ts [hunk 1]
```diff
@@ -61,6 +61,20 @@ export default {
...
+      externalUrl: get(
+        'HMPPS_AUTHORIZATION_SERVER_EXTERNAL_URL',
+        get('HMPPS_AUTHORIZATION_SERVER_URL', 'http://localhost:8089'),
+      ),
+      timeout: {
+        response: Number(get('HMPPS_AUTHORIZATION_SERVER_RESPONSE', 10000)),
+        deadline: Number(get('HMPPS_AUTHORIZATION_SERVER_DEADLINE', 10000)),
+      },
+      agent: new AgentConfig(Number(get('HMPPS_AUTHORIZATION_SERVER_RESPONSE', 10000))),
+      apiClientId: get('API_CLIENT_ID', 'test-client', requiredInProduction),
```
Comment (brightonsbox) [inline server/config.ts:75]: Don't need apiClientId and apiClientSecret defining here.  We'll still be getting authorization_code token from legacy auth for now

File: server/data/baseClientApiClient.test.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  describe('listBaseClients', () => {
+    it('Should return data from the API', async () => {
+      // Given the network is mocked to return a response
+      const testResponse = listBaseClientResponseFactory.build()
+      mockSuccessfulBaseClientRestApiCall(`/base-clients`, testResponse)
+
+      // When we call the API client
+      const promise = baseClientApiClient.listBaseClients()
```
Comment (brightonsbox) [inline server/data/baseClientApiClient.test.ts:54]: Make one line? Just as simple?

`const output = await baseClientApiClient.listBaseClients()`

File: server/data/baseClientApiClient.test.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+  describe('updateBaseClient', () => {
+    it('Should return a success response from the API', async () => {
+      // Given the network is mocked to return a response
+      const testRequest = updateBaseClientRequest.build()
+      mockBaseClientRestApiPutCall(`/base-clients/base_client_id`, 200, null)
+
+      // When we call the API client
+      try {
+        await baseClientApiClient.updateBaseClient('base_client_id', testRequest)
```
Comment (brightonsbox) [inline server/data/baseClientApiClient.test.ts:117]:
```
await expect(baseClientApiClient.updateBaseClient('base_client_id', testRequest))
.resolves.toHaveProperty('status', 200)
```

File: server/data/baseClientApiClient.test.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+      }
+    })
+
+    it('Errors if unsuccessful', async () => {
+      // Given the network is mocked to return an error status code
+      const testRequest = updateBaseClientRequest.build()
+      mockBaseClientRestApiPutCall(`/base-clients/base_client_id`, 400, null)
+
+      // When we call the API client
+      try {
```
Comment (brightonsbox) [inline server/data/baseClientApiClient.test.ts:129]: await expect(baseClientApiClient.updateBaseClient('base_client_id', testRequest))
.rejects.toHaveProperty('status', 400)

File: server/data/baseClientApiClient.test.ts (hunk 2 above)
Comment (thomasridd) [reply server/data/baseClientApiClient.test.ts:54]: yep - extracted to two lines to aid debugging - will replace

File: server/routes/baseClientRouter.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { RequestHandler, Router } from 'express'
+import { Services } from '../services'
+import asyncMiddleware from '../middleware/asyncMiddleware'
+import BaseClientController from '../controllers/baseClientController'
+
+export default function baseClientRouter(services: Services): Router {
+  const router = Router()
+
+  const get = (path: string | string[], ...handlers: RequestHandler[]) =>
```
Comment (brightonsbox) [inline server/routes/baseClientRouter.ts:9]: Any way of lifting these get post etc definitions out of individual router files so they can be reused?

Comment (brightonsbox) [reply server/routes/baseClientRouter.ts:9]: Hmm maybe not unless you pass router in, ignore me

File: server/services/baseClientService.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    const baseClientApiClient = this.baseClientApiClientFactory(token)
+    const response = await baseClientApiClient.listClientInstances(baseClient.baseClientId)
+    return mapListClientInstancesResponse(baseClient, response)
+  }
+
+  async deleteClientInstance(token: string, client: Client): Promise<Response> {
+    const baseClientApiClient = this.baseClientApiClientFactory(token)
+    return baseClientApiClient.deleteClientInstance(client.baseClientId, client.clientId)
+  }
+}
```
Comment (brightonsbox) [inline server/services/baseClientService.ts:65]: This becomes really nice and neat 🙌

File: server/views/partials/list_filter.njk [hunk 7]
```diff
@@ -0,0 +1,86 @@
+{% call govukFieldset({
+legend: {
+text: "Filter clients",
```
Comment (brightonsbox) [inline server/views/partials/list_filter.njk:3]: Indentation of this file looks a little strange?

File: server/data/baseClientApiClient.test.ts (hunk 4 above)
Comment (thomasridd) [reply server/data/baseClientApiClient.test.ts:129]: see above

File: server/services/baseClientService.ts (hunk 6 above)
Comment (thomasridd) [reply server/services/baseClientService.ts:65]: thanks

File: server/routes/baseClientRouter.ts (hunk 5 above)
Comment (thomasridd) [reply server/routes/baseClientRouter.ts:9]: yes - I can take a pattern from approved-premises-ui but I'd like to get this in and create a separate ticket

File: server/views/partials/list_filter.njk (hunk 7 above)
Comment (thomasridd) [reply server/views/partials/list_filter.njk:3]: updated

**Review by brightonsbox — APPROVED** [review]
LGTM

**Pull request merged**